File: backend/api/library.py
from fastapi import APIRouter, HTTPException, UploadFile, File
from fastapi.responses import PlainTextResponse
from pydantic import BaseModel
from typing import Optional, Union
from datetime import datetime
import json

from backend.services.project_manager import get_project_manager
from backend.services.library_manager import (
    get_library_manager,
    DocEntry,
    DocSource,
    DocStatus,
)

router = APIRouter()


class CreateDocRequest(BaseModel):
    name: str
    layer: str
    content: Union[dict, str]
    source: Optional[str] = "manual"
    parent_uid: Optional[str] = None
    directory: str = "/"
    tags: list[str] = []


class UpdateDocRequest(BaseModel):
    name: Optional[str] = None
    directory: Optional[str] = None
    tags: Optional[list[str]] = None
    status: Optional[str] = None


class UpdateContentRequest(BaseModel):
    content: Union[dict, str]


class SetActiveRequest(BaseModel):
    uid: str


class CreateDirectoryRequest(BaseModel):
    path: str


@router.get("/projects/{project_id}/library")
async def list_library(project_id: str, include_archived: bool = False):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    tree = library.get_tree(include_archived=include_archived)
    directories = library.manifest.directories
    
    return {
        "directories": directories,
        "tree": tree,
        "documents": [e.model_dump() for e in library.list_documents(include_archived=include_archived)],
        "active_docs": library.manifest.active_docs,
    }


@router.get("/projects/{project_id}/library/{uid}")
async def get_document(project_id: str, uid: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    result = library.get_document(uid)
    
    if not result:
        raise HTTPException(status_code=404, detail="Document not found")
    
    entry, content = result
    children = library.get_children(uid)
    provenance = library.get_provenance_chain(uid)
    
    return {
        "entry": entry.model_dump(),
        "content": content,
        "children": [c.model_dump() for c in children],
        "provenance": [p.model_dump() for p in provenance],
    }


@router.post("/projects/{project_id}/library")
async def create_document(project_id: str, data: CreateDocRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    source_map = {
        "generate": DocSource.GENERATE,
        "import": DocSource.IMPORT,
        "manual": DocSource.MANUAL,
    }
    source = source_map.get(data.source, DocSource.MANUAL)
    
    uid = library.add_document(
        name=data.name,
        layer=data.layer,
        content=data.content,
        source=source,
        parent_uid=data.parent_uid,
        directory=data.directory,
        tags=data.tags,
    )
    
    return {"uid": uid, "entry": library.get_entry(uid).model_dump()}


@router.put("/projects/{project_id}/library/{uid}")
async def update_document(project_id: str, uid: str, data: UpdateDocRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    update_data = {}
    if data.name is not None:
        update_data["name"] = data.name
    if data.directory is not None:
        update_data["directory"] = data.directory
    if data.tags is not None:
        update_data["tags"] = data.tags
    if data.status is not None:
        update_data["status"] = DocStatus(data.status)
    
    if not update_data:
        raise HTTPException(status_code=400, detail="No fields to update")
    
    entry = library.update_entry(uid, **update_data)
    if not entry:
        raise HTTPException(status_code=404, detail="Document not found")
    
    return {"status": "updated", "entry": entry.model_dump()}


@router.put("/projects/{project_id}/library/{uid}/content")
async def update_content(project_id: str, uid: str, data: UpdateContentRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    success = library.update_content(uid, data.content)
    if not success:
        raise HTTPException(status_code=404, detail="Document not found")
    
    entry = library.get_entry(uid)
    return {"status": "updated", "entry": entry.model_dump()}


@router.delete("/projects/{project_id}/library/{uid}")
async def delete_document(project_id: str, uid: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    success = library.delete_document(uid)
    if not success:
        raise HTTPException(status_code=404, detail="Document not found")
    
    return {"status": "deleted"}


@router.put("/projects/{project_id}/library/{uid}/archive")
async def archive_document(project_id: str, uid: str, archive: bool = True):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    entry = library.archive_document(uid, archive)
    if not entry:
        raise HTTPException(status_code=404, detail="Document not found")
    
    return {"status": "archived" if archive else "unarchived", "entry": entry.model_dump()}


@router.get("/projects/{project_id}/library/active/{layer}")
async def get_active_document(project_id: str, layer: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    result = library.get_active_document(layer)
    
    if not result:
        return {"uid": None, "entry": None, "content": None}
    
    entry, content = result
    return {"uid": entry.uid, "entry": entry.model_dump(), "content": content}


@router.put("/projects/{project_id}/library/active/{layer}")
async def set_active_document(project_id: str, layer: str, data: SetActiveRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    success = library.set_active(layer, data.uid)
    if not success:
        raise HTTPException(status_code=400, detail="Cannot set active: document not found or layer mismatch")
    
    return {"status": "set", "layer": layer, "uid": data.uid}


@router.post("/projects/{project_id}/library/import")
async def import_file(project_id: str, file: UploadFile = File(...)):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    content = await file.read()
    filename = file.filename or "imported_file"
    
    format = "json" if filename.endswith(".json") else "md" if filename.endswith(".md") else "txt"
    
    try:
        if format == "json":
            parsed_content = json.loads(content.decode("utf-8"))
        else:
            parsed_content = content.decode("utf-8")
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Failed to parse file: {str(e)}")
    
    uid = library.import_file(
        name=f"导入 — {filename}",
        content=parsed_content,
        format=format,
    )
    
    return {"uid": uid, "entry": library.get_entry(uid).model_dump()}


@router.post("/projects/{project_id}/library/directories")
async def create_directory(project_id: str, data: CreateDirectoryRequest):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    success = library.create_directory(data.path)
    if not success:
        return {"status": "exists", "path": data.path}
    
    return {"status": "created", "path": data.path}


@router.delete("/projects/{project_id}/library/directories/{path:path}")
async def delete_directory(project_id: str, path: str):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    
    dir_path = "/" + path
    success = library.delete_directory(dir_path)
    if not success:
        raise HTTPException(status_code=400, detail="Cannot delete root or non-existent directory")
    
    return {"status": "deleted", "path": dir_path}


@router.get("/projects/{project_id}/library/export/{uid}")
async def export_document(project_id: str, uid: str, format: str = "json"):
    pm = get_project_manager()
    project = pm.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    library = get_library_manager(project.project_path)
    result = library.get_document(uid)
    
    if not result:
        raise HTTPException(status_code=404, detail="Document not found")
    
    entry, content = result
    
    if format == "txt":
        if isinstance(content, dict):
            content = json.dumps(content, ensure_ascii=False, indent=2)
        return PlainTextResponse(content=content, media_type="text/plain")
    else:
        return {"entry": entry.model_dump(), "content": content}

File: backend-node/src/routes/library.ts
// 文档库路由
import { Express, Request, Response } from 'express';
import { getProjectManager } from '../services/project-manager';
import { getLibraryManager } from '../services/library';
import { DocSource, DocStatus } from '../protocols';

export function registerLibraryRoutes(app: Express): void {
  const pm = getProjectManager();

  function getLibrary(projectId: string) {
    const project = pm.getProject(projectId);
    if (!project) {
      throw new Error('Project not found');
    }
    return getLibraryManager(project.path);
  }

  // 文档树
  app.get('/api/projects/:projectId/library', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const includeArchived = req.query.includeArchived === 'true';
      const tree = library.getTree(includeArchived);
      res.json({ tree });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 新增文档
  app.post('/api/projects/:projectId/library', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const { name, layer, content, source, parentUid, directory, tags } = req.body;
      
      const uid = library.addDocument(name, layer, content, {
        source: source as DocSource,
        parentUid,
        directory,
        tags
      });

      res.json({ uid, status: 'created' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 获取文档
  app.get('/api/projects/:projectId/library/:uid', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const result = library.getDocument(req.params.uid);
      
      if (!result) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json(result);
    } catch (error) {
      res.status(404).json({ error: String(error) });
    }
  });

  // 更新文档
  app.put('/api/projects/:projectId/library/:uid', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const updated = library.updateEntry(req.params.uid, req.body);
      
      if (!updated) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json({ status: 'updated', entry: updated });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 删除文档
  app.delete('/api/projects/:projectId/library/:uid', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const success = library.deleteDocument(req.params.uid);
      
      if (!success) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json({ status: 'deleted' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 更新内容
  app.put('/api/projects/:projectId/library/:uid/content', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const success = library.updateContent(req.params.uid, req.body.content);
      
      if (!success) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json({ status: 'updated' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 归档
  app.put('/api/projects/:projectId/library/:uid/archive', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const archive = req.body.archive !== false;
      const entry = library.archiveDocument(req.params.uid, archive);
      
      if (!entry) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json({ status: archive ? 'archived' : 'unarchived', entry });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 导出
  app.get('/api/projects/:projectId/library/:uid/export', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const result = library.getDocument(req.params.uid);
      
      if (!result) {
        return res.status(404).json({ error: 'Document not found' });
      }

      res.json(result);
    } catch (error) {
      res.status(404).json({ error: String(error) });
    }
  });

  // 活跃文档
  app.get('/api/projects/:projectId/library/active/:layer', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const result = library.getActiveDocument(req.params.layer);
      
      if (!result) {
        return res.status(404).json({ error: 'No active document for this layer' });
      }

      res.json(result);
    } catch (error) {
      res.status(404).json({ error: String(error) });
    }
  });

  app.put('/api/projects/:projectId/library/active/:layer', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const { uid } = req.body;
      const success = library.setActive(req.params.layer, uid);
      
      if (!success) {
        return res.status(400).json({ error: 'Invalid document or layer' });
      }

      res.json({ status: 'set' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 导入文件
  app.post('/api/projects/:projectId/library/import', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const { name, content, format, directory, tags } = req.body;
      
      const uid = library.importFile(name, content, format, directory, tags);
      res.json({ uid, status: 'imported' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  // 目录管理
  app.post('/api/projects/:projectId/library/directories', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const { path } = req.body;
      const success = library.createDirectory(path);
      res.json({ status: success ? 'created' : 'exists' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });

  app.delete('/api/projects/:projectId/library/directories', (req: Request, res: Response) => {
    try {
      const library = getLibrary(req.params.projectId);
      const { path } = req.body;
      const success = library.deleteDirectory(path);
      res.json({ status: success ? 'deleted' : 'not_found' });
    } catch (error) {
      res.status(500).json({ error: String(error) });
    }
  });
}
